### PR TITLE
Bug: groupby multiindex levels equals rows

### LINKED ISSUE #16843: BUG: number of keys for grouping equals axis length causing incorrect grouping
#### Code Sample, a copy-pastable example if possible

```python
df = pd.DataFrame([['foo','two', 0.371118],['foo','two', 0.765483]],columns=['first','second','one'])
df.set_index(['first','second'], inplace=True)

df.groupby(['first','second']).size()

df.reset_index().groupby(['first','second']).size()
```
#### Problem description
The output of the first groupby size gives me an output of:
```
first     1
second    1
```
#### Expected Output

rather then the expected output i get after i reset the index:
```
first  second
foo    two       2
```
The problem appears when the level of multiindex is equal the number of rows of the grouped dataframe (tested for 2 and 3 rows = levels). If the number of rows is smaller or bigger then the levels of the multiindex the problem does not occur.

#### Output of ``pd.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 2.7.13.final.0
python-bits: 64
OS: Darwin
OS-release: 15.4.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None

pandas: 0.20.1
pytest: 3.0.7
pip: 9.0.1
setuptools: 27.2.0
Cython: 0.25.2
numpy: 1.12.1
scipy: 0.19.0
xarray: None
IPython: 5.3.0
sphinx: 1.5.6
patsy: 0.4.1
dateutil: 2.6.0
pytz: 2017.2
blosc: None
bottleneck: 1.2.1
tables: 3.3.0
numexpr: 2.6.2
feather: None
matplotlib: 2.0.2
openpyxl: 2.4.7
xlrd: 1.0.0
xlwt: 1.2.0
xlsxwriter: 0.9.6
lxml: 3.7.3
bs4: 4.6.0
html5lib: 0.999
sqlalchemy: 1.1.9
pymysql: None
psycopg2: None
jinja2: 2.9.6
s3fs: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by P-Tillmann**
Fix for GH16843 where groubpy gives wrong result when number of
groupedby index+columns is equal to length of axis

 closes #16843
 tests added
 passes ``git diff upstream/master --name-only -- '*.py' | flake8 --diff`` (On Windows, ``git diff upstream/master -u -- "*.py" | flake8 --diff`` might work as an alternative.)
 - [ ] whatsnew entry

### CODE REVIEW

File: pandas/tests/groupby/test_groupby.py [hunk 1]
```diff
@@ -3891,6 +3891,17 @@ def predictions(tool):
         result = df2.groupby('Key').apply(predictions).p1
         tm.assert_series_equal(expected, result)
 
+    def test_gb_key_len_equal_axis_len(self):
+            # GH16843
```
Comment (jreback) [inline pandas/tests/groupby/test_groupby.py]: can you add a 1-liner about what this is testing

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm. pls add a whatsnew note (0.21.0 / groupby section)

**Comment by P-Tillmann** [comment]
Ok, lots red warnings now. I hope you have everything from me now for the PR.

**Comment by jreback** [comment]
can you rebase and update for small comments.

**Review by jreback — APPROVED** [review]

File: pandas/tests/groupby/test_groupby.py [hunk 2]
```diff
@@ -3891,6 +3891,19 @@ def predictions(tool):
...
+            # test ensures that index and column keys are recognized correctly
+            # when number of keys equals axis length of groupby
+            df = pd.DataFrame([['foo', 'bar', 'B', 1],
+                               ['foo', 'bar', 'B', 2],
+                               ['foo', 'baz', 'C', 3]],
+                              columns=['first', 'second', 'third', 'one'])
+            df.set_index(['first', 'second'], inplace=True)
```
Comment (jreback) [inline pandas/tests/groupby/test_groupby.py]: can you change to 

``df = df.set_index(..)``

more idiomatic

File: doc/source/whatsnew/v0.21.0.txt [hunk 3]
```diff
@@ -326,4 +327,4 @@ Categorical
 Other
 ^^^^^
 - Bug in :func:`eval` where the ``inplace`` parameter was being incorrectly handled (:issue:`16732`)
-- Bug in ``.isin()`` in which checking membership in empty ``Series`` objects raised an error (:issue:`16991`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.21.0.txt]: what changed on this line?

**Review by jreback — APPROVED** [review]
minor comment. ping when pushed / green.

File: pandas/tests/groupby/test_groupby.py (hunk 2 above)
Comment (gfyoung) [inline pandas/tests/groupby/test_groupby.py]: Also, don't pass in `inplace=True`

**Comment by P-Tillmann** [comment]
@jreback ping ;-)

Would you discourage the use of inplace option in general in favor of assigning back to the df?

**Comment by gfyoung** [comment]
> Would you discourage the use of inplace option in general in favor of assigning back to the df?

Yes, see #16529.

File: doc/source/whatsnew/v0.21.0.txt [hunk 4]
```diff
@@ -296,6 +296,7 @@ Groupby/Resample/Rolling
 - Bug in :func:`infer_freq` causing indices with 2-day gaps during the working week to be wrongly inferred as business daily (:issue:`16624`)
 - Bug in ``.rolling(...).quantile()`` which incorrectly used different defaults than :func:`Series.quantile()` and :func:`DataFrame.quantile()` (:issue:`9413`, :issue:`16211`)
 - Bug in ``groupby.transform()`` that would coerce boolean dtypes back to float (:issue:`16875`)
+- Bug in ``DataFrame.groupby`` when called with index or index + column key numbers equal to the axis length of the groupby (:issue:`16859`)
```
Comment (gfyoung) [inline doc/source/whatsnew/v0.21.0.txt]: Your description underneath your test I think is clearer about the bug than this description is.  Try incorporating some of your test description into here.

**Comment by jreback** [comment]
small changes suggest by @gfyoung look fine. otherwise lgtm.

**Comment by gfyoung** [comment]
@jreback @jorisvandenbossche : Addressed my own comments.  PTAL.

**Comment by gfyoung** [comment]
@jreback @jorisvandenbossche : Addressed my own comments.  PTAL.

**Comment by jreback** [comment]
thanks!

**Pull request merged**